rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+*   Resolvers and importers would only resolve Artifacts produced with same or
+    older versions of the library, by default.
+
+#### SDK (aka DSL), Official Components {#sdk-aka-dsl-official-components}
+
+All parts of SDK (aka DSL), IR, interfaces of official components shall comply
+with SemVer. The functionalities form the following core CUJs, which include but
+are not limited to:
+
+*   Defining or declaring components;
+*   Composing them into a pipeline graph;
+*   Environment/Deployment specific decoration of nodes;
+*   Advanced control semantics (condition/loop).
+
+It’s worth noting again that APIs from underlying packages used in above core
+CUJs shall also comply.
+
+#### Custom Component {#custom-component}
+
+There are several ways to define a Custom Component in TFX::OSS:
+
+1.  Importing a user provided python function
+    *   this process and all APIs involved will be backward compatible;
+2.  Importing a user provided container image
+    *   this process and APIs will be backward compatible;
+3.  Defining ComponentSpec, ExecutorSpec and subclassing with component class
+    *   This is a very complex user interface, and arguably exposes too much
+        internal implementation details. Moreover, it’s proven that this is
+        quite difficult to do for our users. A possible improvement proposed is
+        to combine the component and executor classes, and have the convention
+        be that users subclass a base component's class to extend or override
+        functionality. Until this is settled, this API will not comply with
+        backward compatibility.
+
+#### Customization on ExampleGen {#customization-on-examplegen}
+
+There are several ways to customize ExampleGen:
+
+1.  Providing a PTransform for converting input source to examples by inheriting
+    from BaseExampleGenExecutor and overwriting the abstract method.
+    *   This is recommended and will be backward compatible.
+2.  Use the above custom Component or custom Executor for use cases like join.
+    *   This is a very complex user interface, and arguably exposes too much
+        internal implementation details (driver:external file detection,
+        span/version/split etc). Moreover, it’s proven that this is quite
+        difficult to do for our users. We plan to mark this process as
+        experimental in 1.0, and not comply with backward compatibility.
+
+#### Custom Executor {#custom-executor}
+
+Currently TFX supports “custom executor”, but essentially this means providing a
+different executor to an official TFX component. So far we have found two
+concrete use cases:
+
+1.  Google-cloud optimized tuner and trainer;
+2.  Customizing ExampleGen component.
+
+Because this process requires users to understand many details of internal
+implementations of an official TFX component, we plan to deprecate this support
+externally and declare it internal only. We will only support currently known
+use cases listed above.
```

Comment:
What would then be the replacement if we want to customize the execution logic for Trainer (ex: launching a K8s based trainer on a custom pod). Would there be some way to customize this or would it require overwriting the Trainer component (and therefore giving up backwards compatibility)

Reply:
If you are customizing a trainer and launching k8s pod, I feel that a Python or container based custom component would not be much harder than fitting a custom executor in place of our Trainer component. And those two experiences should receive more support.

If you must use this path and design your executor to take the same conventions as the CAIP executors, you might receive some warnings saying you are using an unsupported executor class, but we have no plan to remove this functionality so your code would still work post 1.0.

So, if you already have this executor, we could try to make it work. But if it's undeveloped yet, I would recommend you to use Python or container based solution. WDYT?

@1025KB to double check above promise can be preserved.

Reply:
>So, if you already have this executor, we could try to make it work. But if it's undeveloped yet, I would recommend you to use Python or container based solution. WDYT?

It's currently undeveloped, will take the guidelines as pointed here ty!